### DISCUSSION TITLE
Dependency Injection Freezes Startup of Application

### DISCUSSION THREAD

**Discussion by Spaxterr** [quartznet/quartznet#2152] [Q&A] [answered]
Hello, I'm having some issues that seems to have with DI to do. I have a class that implements `SchedulerListenerSupport` which looks something like this.

```csharp
public class DefaultScheduleListener : SchedulerListenerSupport
{
    private readonly ILogger<DefaultScheduleListener> _logger;
    private readonly IScheduler _scheduler;
    private readonly SchedulerDbContext _dbContext;

    public DefaultScheduleListener(ILogger<DefaultScheduleListener> logger, SchedulerDbContext schedulerDbContext, IScheduler scheduler)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _dbContext = schedulerDbContext ?? throw new ArgumentNullException(nameof(schedulerDbContext));
        _scheduler = scheduler ?? throw new ArgumentNullException(nameof(scheduler));
    }
}
```

However, it seems like trying to inject `IScheduler` causes my application to freeze half-way through the startup process. I see the following logs before the app just freezes and won't do anything until I stop it.
![image](https://github.com/quartznet/quartznet/assets/39764099/633d9e5a-9b7d-47d9-a989-aa799bbaf0ff)

I am unsure of what to do about this, I have checked for circular dependencies and found none. By using a debugger I've been able to determine that the constructor of this class does not even run before the app freezes. When I remove the `IScheduler` parameter, the app will run like usual. Are there any other methods of accessing an `IScheduler` from this class?

**Comment by lahma** [comment, marked as answer]
You need to inject `ISchedulerFactory` and then resolve the `IScheduler` when needed (outside of constructor) with `var scheduler = await schedulerFactory.GetScheduler()`.

**Reply by Spaxterr** [reply]
Thanks, that works. However it now seems like the `UseMicrosoftDependencyInjectionJobFactory` no longer works as whenever I try to schedule a new job the following error is thrown.

```
Quartz.SchedulerException: Problem instantiating class 'scheduler.Job.CustomJob: Cannot instantiate type which has no empty constructor (Parameter 'CustomJob')'
```

**Reply by lahma** [reply]
Check ASP.NET Core example: https://github.com/quartznet/quartznet/tree/3.x/src/Quartz.Examples.AspNetCore, it should work (and without `UseMicrosoftDependencyInjectionJobFactory `), you should use the latest 3.7 release.

**Reply by Spaxterr** [reply]
Right, I noticed that it was deprecated so I removed it. I do not see how they are using `ISchedulerFactory` in their example, but other than that my implementation looks pretty much the same. Correct me if I have got the wrong idea with my DI.

`Startup.cs`
```C#
services.
// ...
.AddSingleton<ISchedulerFactory, StdSchedulerFactory>()
.AddTransient<MyCustomJob>()
// ...
```

`DefaultScheduleListener.cs`
```C#
    private readonly ILogger<DefaultScheduleListener> _logger;
    private readonly SchedulerDbContext _dbContext;
    private readonly ISchedulerFactory _schedulerFactory;

    public DefaultScheduleListener(ILogger<DefaultScheduleListener> logger, IDispatcher dispatcher, SchedulerDbContext schedulerDbContext, ISchedulerFactory schedulerFactory)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        _dbContext = schedulerDbContext ?? throw new ArgumentNullException(nameof(schedulerDbContext));
        _schedulerFactory = schedulerFactory ?? throw new ArgumentNullException(nameof(schedulerFactory));
    }
    
    public async void SomeMethod()
    {
        var scheduler = await _schedulerFactory.GetScheduler();
    }
```

Reading the stacktrace, it seems like it's using the `SimpleJobFactory`, is this correct?
![image](https://github.com/quartznet/quartznet/assets/39764099/e4d29105-116c-4965-9273-923dbbd31278)

**Reply by lahma** [reply]
You shouldn't register scheduler factory yourself. You should use `AddQuartz()` like in the example.
